Fix of Issues #130 and #83

dongbohu opened the pull request:
This PR is filed to fix issue #130 (and #83 as a byproduct) in **Signature** (aka. Node) page. I touched more code than expected:

1. Removed the samples in each experiment.

2. Made the vega-lite plot clickable. When it is clicked, the page will go to annotations of ALL samples in current experiment.

3. Refactored all resource classes into `resources.js`.

4. Used `$resource` to replace `$http` service.

5. Moved the code of vega-lite specification out of `highRangeExp` directive, and created a new service *EmbedSpecService* for it. It makes the code more readable.

6. Added "data_source" field in sample annotations web UI.

7. Removed redundant resource class in `sampleBin.js`, which refers to `signature/resources.js`.

8. Tweaked the API paths in signature module's test cases.

dongbohu commented:
Demo site:
http://165.123.67.202/#/signature/11344

dongbohu commented on interface/src/app/signature/signature.spec.js:
```diff
@@ -153,27 +153,27 @@ describe('<high-range-exp> directive', function() {
-      {id: 19, signature: 123, sample: 1019, value: 0.0661371872868119},
-      {id: 20, signature: 123, sample: 1020, value: 0.0783672660119669},
-      {id: 21, signature: 123, sample: 1021, value: 0.0769023791604038}
+      {id: 1, node: 123, sample: 1001, value: 0.0854334209211516},
```
I changed the field name in this test case back to `node`, because adage's back end is still using `node` instead of `signature` as field name. We will change these fields when `node` field at the back end is renamed to `signature`.

rzelayafavila replied:
:+1:

dongbohu commented on interface/src/app/signature/resources.js:
```diff
@@ -13,6 +13,19 @@ angular.module('adage.signature.resources', [
 .factory('Participation', ['$resource', 'ApiBasePath',
   function($resource, ApiBasePath) {
-    return $resource(ApiBasePath + 'participation/');
+    return $resource(ApiBasePath + 'participation');
```
By default the trailing slashes will be removed. So I removed it in the test cases to make it consistent.

rzelayafavila replied:
Ok

dongbohu commented on interface/src/app/signature/signature.js:
```diff
@@ -92,8 +92,71 @@ angular.module('adage.signature', [
+  sample2index = {};
+
+  this.mode = 'vega-lite';
+  this.actions = false;
```
`mode` and `actions` are properties only required by the vega-lite specification.

rzelayafavila replied:
Oh ok, I see :+1:

dongbohu commented:
@rzelayafavila: About your question on the location of the link to sample annotations page, it is a good question. I feel that since the sample annotations page is related to the samples (the red lines in the vega-lite activity distribution plot), it makes more sense to add the link to the vega-lite plot (instead of to the experiment name). Does this make sense?

dongbohu commented:
On the other hand, if the URL points to a page of experiment information, then the experiment name will be definitely a better place to put the link onto.

rzelayafavila commented:
@dongbohu - In response to your first answer, are the samples always the same for any given experiment on the next page? If so, maybe we can add it to the parent element that contains both the vega-lite plot and the experiment name (this is only if both are always linked to the same page)?

dongbohu commented:
@rzelayafavila: I added the tooltip for the vega-lite widget. Please take a look at the code and demo site and let me know what you think. Thanks. 

Just so you know: I am also working on the workflow redesign, so some of the other demo site pages (such as signature search) is not working now.

rzelayafavila commented:
Hi @dongbohu - I checked it out and it looks good! Do you know why the tests are failing in CircleCI though?

rzelayafavila commented:
Hmmm, it looks like it is complaining about the urllib3 version.

dongbohu commented:
Yeah, I remembered that too.

dongbohu commented:
@rzelayafavila: I updated `requirements.txt`, also added some comments in `high_range_exp.tpl.html` to explain why the tooltip is placed on the left side instead of right side. Please let me know if you have any other concerns.
